File: src/main/java/com/claymccoy/jcalc/CalculatorImpl.java
package com.claymccoy.jcalc;

import com.claymccoy.jcalc.model.Calculator;
import com.claymccoy.jcalc.model.Expression;
import jakarta.validation.constraints.NotNull;

import static java.util.Objects.requireNonNull;

final public class CalculatorImpl implements Calculator
{
    @Override
    public int calculate(@NotNull Expression expression)
    {
        requireNonNull(expression, "expression is null");
        return switch (expression.operator()) {
            case Add -> expression.first().value() + expression.second().value();
            case Subtract -> expression.first().value() - expression.second().value();
            case Multiply -> expression.first().value() * expression.second().value();
            case Divide -> expression.first().value() / expression.second().value();
            case Exponent -> calculateExp(expression.first().value(), expression.second().value());
        };
    }

    private int calculateExp(int base, int exponent) {
        var acc = 1;
        for (int i = exponent; i > 0; i--) {
            acc = acc * base;
        }
        return acc;
    }
}

File: src/main/kotlin/com/claymccoy/kcalc/CalculatorImpl.kt
package com.claymccoy.kcalc

import com.claymccoy.kcalc.Operator.*

class CalculatorImpl: Calculator {
    override fun calculate(expression: Expression) : Int {
        return when (expression.operator) {
            Add -> expression.run { first.value + second.value }
            Subtract ->  expression.run { first.value - second.value }
            Multiply ->  expression.run { first.value * second.value }
            Divide ->  expression.run { first.value / second.value }
            Exponent -> expression.run { calculateExp(first.value, second.value) }
        }
    }

    private tailrec fun calculateExp(base: Int, exponent: Int, acc: Int = 1): Int {
        if (exponent == 0) return acc
        return calculateExp(base, exponent - 1, acc * base)
    }
}
